alerts/actions/triage_bot.py

```diff
@@ -0,0 +1,746 @@
+                else:
+                    logger.error("Updating duplicate chain")
+                    operation = _update_duplicate_chain
+                    should_dispatch = False
+
+                operation(
+                    self._rest_api_cfg,
+                    request.alert,
+                    request.user,
+                    [request.identifier],
+                )
+            except APIError as err:
+                # In the case that we fail to maintain a duplicate chain,
+                # we should default to messaging users even at the risk of being
+                # noisy, as doing so is a useful indication of failure.
+                logger.exception(
+                    "Duplicate chain management error: {}".format(err.message),
+                )
+                should_dispatch = True
+
+            if should_dispatch:
+                result = dispatch(request, self._lambda_function_name)
+
+                # In the case that dispatch fails, attempt to re-discover the name
+                # of the lambda function to invoke in case it was replaced.
+                if result != DispatchResult.SUCCESS:
+                    logger.error("Failed to dispatch request")
+                    reset = timedelta(seconds=L_FN_NAME_VALIDITY_WINDOW_SECONDS)
+                    self._last_discovery = datetime.now() - reset
+
+        return alert
+
+    def _oauth_handshake(self):
+        self._person_api_session = authenticate(
+            OAUTH_URL,
+            AuthParams(
+                client_id=self._config["person_api_client_id"],
+                client_secret=self._config["person_api_client_secret"],
+                audience=PERSON_API_AUDIENCE,
+                scope=PERSON_API_SCOPE,
+                grants=PERSON_API_GRANTS,
+            ),
+        )
+
+        if self._person_api_session is None:
+            logger.Exception("Failed to establish OAuth session")
+            raise AuthFailure()
+
+        self._last_authenticated = datetime.now()
+
+    def _discover_lambda_fn(self):
+        functions = [
+            function
+            for function in _discovery(self._boto_session)()
+            if function.name.startswith(SLACK_BOT_FUNCTION_NAME_PREFIX)
+        ]
+
+        if len(functions) == 0:
+            logger.Exception("Failed to discover Lambda function")
+            raise DiscoveryFailure()
+
+        self._lambda_function_name = functions[0].name
+
+        self._last_discovery = datetime.now()
+
+
+def try_make_outbound(
+    alert: Alert, oauth_tkn: Token
+) -> types.Optional[AlertTriageRequest]:
+    """Attempt to determine the kind of alert contained in `alert` in
+    order to produce an `AlertTriageRequest` destined for the web server comp.
+    """
+
+    _source = alert.get("_source", {})
+    category = _source.get("category")
+    tags = _source.get("tags", [])
+
+    is_sensitive_host_access = "session" in tags and category == "session"
```

Comment:
Is there a way that instead of hardcoding which alerts we handle in the code, can we abstract that away to the config file?

Reply:
I believe we talked about this a while ago and concluded that there's no generic way of handling this with data alone.  Each alert that the bot is able to process has to follow some distinct steps to retrieve information about a user (duo alerts provide their email directly while others require a query to the PersonAPI for example). So, at some level, we always need a code addition to map some identifying information in an alert to a function that can process the information in that alert and collect additional information necessary to triage it.